File: vespalib/src/vespa/vespalib/trace/slime_trace_deserializer.cpp
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
#include <vespa/vespalib/trace/slime_trace_deserializer.h>
#include <vespa/vespalib/trace/slime_trace_serializer.h>
#include <vespa/vespalib/trace/tracenode.h>

using namespace vespalib::slime;

namespace vespalib {

SlimeTraceDeserializer::SlimeTraceDeserializer(const Inspector & inspector)
    : _inspector(inspector)
{
}

TraceNode
SlimeTraceDeserializer::deserialize() const
{
    return deserialize(_inspector);
}

TraceNode
SlimeTraceDeserializer::deserialize(const Inspector & inspector)
{
    TraceNode node(deserializeTraceNode(inspector));
    deserializeChildren(inspector[SlimeTraceSerializer::CHILDREN], node);
    return node;
}

TraceNode
SlimeTraceDeserializer::deserializeTraceNode(const Inspector & inspector)
{
    system_time timestamp(std::chrono::milliseconds(decodeTimestamp(inspector)));
    if (hasPayload(inspector)) {
        std::string note(decodePayload(inspector));
        return TraceNode(note, timestamp);
    }
    return TraceNode(timestamp);
}

bool
SlimeTraceDeserializer::hasPayload(const Inspector & inspector)
{
    return inspector[SlimeTraceSerializer::PAYLOAD].valid();
}

vespalib::string
SlimeTraceDeserializer::decodePayload(const Inspector & inspector)
{
    return inspector[SlimeTraceSerializer::PAYLOAD].asString().make_string();
}

int64_t
SlimeTraceDeserializer::decodeTimestamp(const Inspector & inspector)
{
    return inspector[SlimeTraceSerializer::TIMESTAMP].asLong();
}

void
SlimeTraceDeserializer::deserializeChildren(const Inspector & inspector, TraceNode & node)
{
    for (size_t i(0); i < inspector.children(); i++) {
        Inspector & child(inspector[i]);
        TraceNode childNode(deserialize(child));
        node.addChild(childNode);
    }
}

}

File: config/src/main/java/com/yahoo/vespa/config/protocol/SlimeTraceDeserializer.java
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.vespa.config.protocol;

import com.yahoo.slime.ArrayTraverser;
import com.yahoo.slime.Inspector;
import com.yahoo.yolean.trace.TraceNode;

/**
 * Deserializing from a {@link Inspector} (slime) representation to a {@link TraceNode}
 *
 * @author Ulf Lilleengen
 */
public class SlimeTraceDeserializer {
    private final Inspector entry;
    public SlimeTraceDeserializer(Inspector inspector) {
        this.entry = inspector;
    }

    public TraceNode deserialize() {
        return deserialize(entry);
    }

    private static TraceNode deserialize(Inspector entry) {
        Object payload = decodePayload(entry.field(SlimeTraceSerializer.PAYLOAD));
        long timestamp = decodeTimestamp(entry.field(SlimeTraceSerializer.TIMESTAMP));
        final TraceNode node = new TraceNode(payload, timestamp);
        Inspector children = entry.field(SlimeTraceSerializer.CHILDREN);
        children.traverse(new ArrayTraverser() {
            @Override
            public void entry(int idx, Inspector inspector) {
                node.add(deserialize(inspector));
            }
        });
        return node;
    }

    private static long decodeTimestamp(Inspector entry) {
        return entry.asLong();
    }

    private static Object decodePayload(Inspector entry) {
        return switch (entry.type()) {
            case STRING -> entry.asString();
            case LONG -> entry.asLong();
            case BOOL -> entry.asBool();
            case DOUBLE -> entry.asDouble();
            case DATA -> entry.asData();
            default -> null;
        };
    }
}
